[t-leongchuan] iP

t-leongchuan opened the pull request:
# Denim 👖 
> "The name is inspired by a certain dream." – Me

Denim _sorta_ helps you but not as well as how a phone might. But hey, it:
- Spit cute responses when you enter the wrong command
- Manages your tasks such as deadlines, events and to-dos. It is too lazy to bother reminding you though. 😪 
- has probably some bugs that need to be squashed. 🐛 

To try it, all you need to so is:

1. Download the JAR file [here](https://github.com/t-leongchuan/ip/releases)
2. Run the JAR file in your terminal
3. Start using it

To start, let it spit out the possible commands and format you need to use by entering the command
`help`

Have fun using it, and use it as a refactoring exercise! I find some part of the code to be unclean 💩 , especially the Abstract `Command` class as well as its subclasses:
``` Java
public abstract class Command {
    public abstract CommandResult execute(TaskList taskList, TaskIo taskIo);
}
```    

Features:
- [x]  Manages Tasks
- [x] GUI (Coming Soon)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

t-leongchuan committed:
Rename Duke to Denim

implement initial chatbot with greeting and exit message.

t-leongchuan committed:
Add support for echo

t-leongchuan committed:
Add support for add and list

t-leongchuan committed:
Add initial Task.java file for managing tasks

t-leongchuan committed:
Add supporting methods for Task.java file

* toString(): Overriden method to format string output
* setDone(): Setter to retrieve boolean isDone

t-leongchuan committed:
Add JavaDoc comments for methods in Task.java

t-leongchuan committed:
Add support for unmarking and marking task

t-leongchuan committed:
Add Deadline.java

Deadline inherits from Task.

t-leongchuan committed:
Add Event.java

Event inherits from Task.

t-leongchuan committed:
Add Todo.java

Todo inherits from Task.

t-leongchuan committed:
Fix string formatting bug in Task.java

t-leongchuan committed:
Refactor Denim and add support for tracking three types of tasks

Tasks Supported:
1.Todo: tasks without any date/time attached to it
2.Deadlines: tasks that need to be done before a specific date/time
3.Events: tasks that start and end at specific date/time

t-leongchuan committed:
Semi-automate testing of Denim

t-leongchuan committed:
Add DenimException.java to handle Denim exceptions

t-leongchuan committed:
Add support for handling basic errors

t-leongchuan committed:
Add support for basic deletion of tasks

t-leongchuan committed:
Refactor Task data structure to use ArrayList

t-leongchuan committed:
Refactor task types to use enums

t-leongchuan committed:
Modify access modifiers for methods to 'private'

t-leongchuan committed:
Add .gitignore file to exclude data directory

t-leongchuan committed:
Add methods to read from and write to data/denim.txt

t-leongchuan committed:
Add additional constructor to Task and its subclasses

t-leongchuan committed:
Refactor Denim.java for better isolation of logic

Isolate I/O operations: Moved all file reading/writing logic to
TaskIO.java to adhere to the single responsibility principle.

Delegate command parsing: Shifted user input parsing and command
routing to TaskManager.java for better separation of concerns.

Manage tasks centrally: Ensured TaskList.java is solely responsible
for managing the collection of Task objects, encapsulating
all ArrayList operations.

t-leongchuan committed:
Add method to Task and subclasses for standardized file I/O

toSimplifiedString() method added to standardize I/O.

t-leongchuan committed:
Add support for reading and writing tasks to data/denim.txt

* Implemented functionality to read tasks from data/denim.txt at
  startup.
* Added functionality to write tasks to data/denim.txt when
  new tasks are added.
* Deletion of tasks and marking/unmarking tasks as complete are not
  yet supported in this commit.

t-leongchuan committed:
Add support for deletion of task to denim.txt

t-leongchuan committed:
Add support of marking and unmarking tasks to denim.txt

t-leongchuan committed:
Enhance user error message for corrupted denim.txt file

t-leongchuan committed:
Merge branch 'branch-Level-7'

t-leongchuan committed:
Enhance UX by displaying a message when the task list is empty

t-leongchuan committed:
Add support for Deadline tasks to recognize dates as LocalDateTime

t-leongchuan committed:
Merge branch 'branch-Level-8'

t-leongchuan committed:
Add command classes and implementations for task management

t-leongchuan committed:
Abstract Parser from TaskManager for better separation of logic

t-leongchuan committed:
Implement command for unmarking task

t-leongchuan committed:
Implement command for adding todo tasks

t-leongchuan committed:
Add command for exit and help

t-leongchuan committed:
Add support for command to read/write task and check errors

t-leongchuan committed:
Remove obsolete TaskManager.java

t-leongchuan committed:
Add support for Event Tasks to use LocalDateTime

t-leongchuan committed:
Add support for Parser to recognize Exit and Help Command

t-leongchuan committed:
Abstract Ui from Denim for better separation of logic

t-leongchuan committed:
Add support for TaskIO to read Event Tasks with LocalDateTIme

t-leongchuan committed:
Add methods to support more ArrayList<Task> operations

t-leongchuan committed:
Refactor Denim to improve program flow and readability

t-leongchuan committed:
Merge branch 'A-MoreOOP'

t-leongchuan committed:
Organize program into respective packages to enhance modularity

t-leongchuan committed:
Merge branch 'A-Packages'

t-leongchuan committed:
Fix index handling bug in marking, unmarking, and delete operations

t-leongchuan committed:
Merge branch 'add-gradle-support'

t-leongchuan committed:
Add Gradle support

t-leongchuan committed:
Fix String formatting bug in TaskList and Commands

t-leongchuan committed:
Add support for testing Parser and TaskList

t-leongchuan committed:
Remove unnecessary imports from Commands and inherited classes

t-leongchuan committed:
Add JavaDoc comments for some files

t-leongchuan committed:
Add Find Command and Support for TaskList to find Task

t-leongchuan committed:
Add support for finding tasks in the task list

t-leongchuan committed:
Adhere Commands to Code Conventions

t-leongchuan committed:
Adhere Tasks and inherited classes to Coding Conventions

t-leongchuan committed:
Fix minor coding convention violations in Commands and Tasks

t-leongchuan committed:
Adhere TaskList to coding conventions

t-leongchuan committed:
Fix minor Coding Convention violation in TaskList

t-leongchuan committed:
Adhere TaskIo to coding conventions

t-leongchuan committed:
Adhere remaining files with Coding Conventions

t-leongchuan committed:
Merge branch 'A-JavaDoc'

t-leongchuan committed:
Merge branch 'A-CodingStandard'

# Conflicts:
#	src/main/java/Denim/Commands/Command.java
#	src/main/java/Denim/Storage/TaskIO.java
#	src/main/java/Denim/TaskList.java

t-leongchuan committed:
Merge branch 'Level-9'

# Conflicts:
#	src/main/java/Denim/Parser.java
#	src/main/java/Denim/TaskList.java

t-leongchuan committed:
Fix minor coding convention violation and fix mainClass in Gradle

woke02 reviewed (commented):
LGTM. Overall, code follows coding standard, OOP is applied appropriately and code is well structured. Well Done!

t-leongchuan committed:
Add checkstyle capability and fix Test files coding conventions

t-leongchuan committed:
Fix coding convention of Command and inherited classes

t-leongchuan committed:
Fix some files with coding style violation via checkstyle

t-leongchuan committed:
fix remaining files with coding violations via checkstyle

t-leongchuan committed:
Add first iteration of GUI

t-leongchuan committed:
Add second iteration of GUI

t-leongchuan committed:
Add iteration 3 of GUI

t-leongchuan committed:
Integrate GUI with core logic

t-leongchuan committed:
Add styling to GUI

t-leongchuan committed:
Merge branch 'A-CheckStyle'

t-leongchuan committed:
Merge branch 'Level-10'

# Conflicts:
#	src/main/java/Denim/Denim.java
#	src/main/java/Denim/Parser.java
#	src/main/java/Denim/TextBasedUi.java

t-leongchuan committed:
Fix remaining inconsistencies in Denim and Parser

t-leongchuan committed:
Fix coding style violating for GUI files

t-leongchuan committed:
Modify FindCommand to accept multiple keywords for finding tasks

t-leongchuan committed:
Merge branch 'A-Varargs'

kimxw commented on src/main/java/Denim/Commands/Command.java:
Great use of packages like Command, Storage, Task etc
To follow the convention, your package names should be all lower case
eg. denim.commands

t-leongchuan replied:
Yup I'm aware it's just that still trying to learn and fiddle around with Git as the suggestion above is only local to my computer but not on GitHub.

As for why I feel that some part is messy, I think one of the better examples is in TaskIo class. Reading through the code, I realized that there were a lot of quick-fixes and separation of concern/logic was subpar for the class. 

Even though the methods are separated out to handle cases (such as file not found, or directory not found), there is still code smell in the sense that the method (handleFileNotFound() and handleDirectoryNotFound()) are still doing way too many things and is even assuming that the user-interface is in the CLI 
(Which of course, with the GUI we just used last week, that is no longer the case.)

Additionally, my initial plan was to create two classes out of TaskIo, where one is for specifically reading tasks and writing tasks. However, I figured that I was probably too lazy then to untangle the mess I made and I am currently trying to refactor the class now.

kimxw commented on src/main/java/Denim/Commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,43 @@
+
+    public DeadlineCommand(String taskDescription, LocalDateTime deadline) {
+        deadlineTask = new Deadline(taskDescription, deadline);
+    }
```
Add a new line between methods. Same issue at some other points.

kimxw commented on src/main/java/Denim/Commands/HelpCommand.java:
```diff
@@ -0,0 +1,31 @@
+
+    public static final String COMMAND_WORD = "help";
+
+    public final String returnFormat = String.format("%s\n%s\n%s\n%s\n%s\n%s\n%s",
```
Great work with the abstraction here. I should learn from this for my own ip.

kimxw commented on src/main/java/Denim/Parser.java:
```diff
@@ -0,0 +1,264 @@
+
+        if (!matcher.matches()) {
+            return new InvalidCommand("Wrong format for delete command", DeleteCommand.COMMAND_USAGE);
+        }
```
new line after egyptian brackets to improve readibility.

kimxw commented on src/main/java/Denim/Storage/TaskIO.java:
```diff
@@ -0,0 +1,275 @@
+     */
+    private void createSavePoint(FileStatus status, Scanner sc) throws DenimException {
+        switch (status) {
+        case DIRECTORY_DOES_NOT_EXIST:
```
It's interesting that you chose to use an enum for file status. While it's definitely good practice, I'm curious about the point at which enums might become less advantageous compared to using else if statements.
Would like to know your opinion

t-leongchuan replied:
Hi @kimxw , the reason why I used `enums` is just for better readability than anything else. I think it is fine to use 
`if else` statement here as there are only a few cases to consider here.

kimxw commented on src/test/java/Denim/ParserTest.java:
Correct directories setup for gradle test ✅

kimxw commented on src/test/java/Denim/ParserTest.java:
```diff
@@ -0,0 +1,94 @@
+
+    private final Parser parser = new Parser();
+    @Test
+    public void prepareEvent_validInput_returnsEventCommand() {
```
test method names follow naming convention ✅

kimxw reviewed (commented):
Honestly great work. Going through your work has given me inspiration to improve my own work. Eventhough you said that the OOP is messy at places, I did not feel that way going through the code. Perhaps because you are more familiar with your own work you feel like things could be improved. But this project is great in my opinion!

t-leongchuan committed:
Update build.gradle to enable Java Assertions

t-leongchuan committed:
Add assertions for input validation

Added assertions in TaskList, DeleteCommand, MarkCommand,
and UnmarkCommand to validate that indices are within the valid
range of the TaskList.

t-leongchuan committed:
Improve code quality through refactoring and comments

- Main: Renamed Duke to Denim in comments
- Parser: Clarified response when user input is unknown
- TaskList: Refactored to replace magic literals
- TaskIo: Added fallthrough comment for case with no break statement

t-leongchuan committed:
Change TaskList findTask() method implementation to use Streams

t-leongchuan committed:
Remove Text-Based UI

The text-based UI is no longer needed as it is replaced with a GUI.

TextBasedUi is being removed and any classes dependent on
or associated with TextBasedUi are refactored to be consistent
with only using the GUI.

Given that the code is mostly obsolete now, it is best to remove it
rather than keep it as we can always retrieve it back if we need it
again later.

t-leongchuan committed:
Fix inconsistencies with coding style and code quality

t-leongchuan committed:
Add new DenimFileException for Exceptions relating to file I/O.

t-leongchuan committed:
Refactor: Extract Class from TaskIo

TaskIo is doing both reading and writing to file.

TaskIo has been removed and two new classes extracted from it,
which are ReadTaskFile and WriteTaskFile.

This is for better separation of logic.

t-leongchuan committed:
Add support for GUI to handle file issues when loading

t-leongchuan committed:
Fix coding style errors

t-leongchuan committed:
Remove TaskIO

t-leongchuan committed:
Merge pull request #1 from t-leongchuan/heads/A-Assertions

Add assertions to TaskList methods

t-leongchuan committed:
Merge branch 'master' into A-CodeQuality

# Conflicts:
#	src/main/java/Denim/Commands/DeleteCommand.java
#	src/main/java/Denim/Commands/MarkCommand.java
#	src/main/java/Denim/Commands/UnmarkCommand.java

t-leongchuan committed:
Merge branch 'master' into A-Streams

t-leongchuan committed:
Merge branch 'A-CodeQuality'

t-leongchuan committed:
Merge branch 'master' into A-Streams

t-leongchuan committed:
Merge branch 'A-Streams'

t-leongchuan committed:
Rename package name to be consistent with coding style

t-leongchuan committed:
Set up CI

t-leongchuan committed:
Merge branch 'A-CI'

t-leongchuan committed:
Fix bug where event start time can be later than event end time

t-leongchuan committed:
Provide in-App guidance to users in the GUI

t-leongchuan committed:
Merge branch 'C-Help'

t-leongchuan committed:
Separate HelpWindow into its own stage

t-leongchuan committed:
Fix coding style violations

t-leongchuan committed:
Add better styling for HelpWindow

t-leongchuan committed:
Add additional styling for different outcome of executing command

t-leongchuan committed:
Add support for program to handle corrupted files

t-leongchuan committed:
Fix event task not being able to be added

t-leongchuan committed:
Merge branch 'A-BetterGui'

t-leongchuan committed:
Merge branch 'A-MoreErrorHandling'

# Conflicts:
#	src/main/java/Denim/Main.java
#	src/main/java/Denim/TaskList.java

t-leongchuan committed:
Fix event task not being read properly

t-leongchuan committed:
Add personalized styling to GUI and reply messages

t-leongchuan committed:
Merge branch 'A-Personality'

t-leongchuan committed:
Fix coding style violations

t-leongchuan committed:
Update README

t-leongchuan committed:
Add additional test for ParserTest

t-leongchuan committed:
Merge branch 'A-MoreTesting'

t-leongchuan committed:
Generate more test with use of AI

t-leongchuan committed:
Fix help command bug

t-leongchuan committed:
Merge branch 'A-AiAssisted'

t-leongchuan committed:
Fix bug with delete command

t-leongchuan committed:
Add User Guide

t-leongchuan committed:
Merge branch 'A-UserGuide'

t-leongchuan committed:
Add additional picture for User Guide

t-leongchuan committed:
Fix bug with program not handling corrupted file gracefully

t-leongchuan committed:
Rename screenshot to Ui.png
